third_party/move/move-vm/runtime/src/interpreter.rs

```diff
@@ -146,19 +154,40 @@ impl Interpreter {
         }
 
         let mut current_frame = self
-            .make_new_frame(gas_meter, loader, function, ty_args, locals)
+            .make_new_frame(
+                gas_meter,
+                entrypoint_resolver.loader(),
+                function,
+                ty_args,
+                locals,
+            )
             .map_err(|err| self.set_location(err))?;
 
         // Access control for the new frame.
         self.access_control
             .enter_function(&current_frame, current_frame.function.as_ref())
             .map_err(|e| self.set_location(e))?;
+
         loop {
-            let resolver = current_frame.resolver(loader, module_store);
-            let exit_code =
-                current_frame //self
-                    .execute_code(&resolver, &mut self, data_store, module_store, gas_meter)
-                    .map_err(|err| self.attach_state_if_invariant_violation(err, &current_frame))?;
+            // Select which resolver to use: if we are executing in the context of the entrypoint
+            // function, we use the one provided by the caller.
+            let resolver = if self.call_stack.is_empty() {
```

Comment:
is this new logic? previously it worked through Script / Module distinction, right? let's be careful here, but it does seem we were creating resolvers every time for scripts and now we re-use? can call stack be cleared e.g. on abort, and then have unintended consequences? just asking.
Maybe we can encapsulate the logic somewhere (and add any other safeguards/checks if needed)?

Reply:
Yes, good point. I rewrote that ugly APIs by storing `LoadedFunction` everywhere, and providing it with scope. @runtian-zhou I think this looks cleaner, and V2 loader API fit perfectly. WDYT? See the last commit in this PR (Note: I did not polish certain types, etc., but added this code to give an idea).

Reply:
Update: this change is refactored into #14185